Docs for describing async compaction and how to operate it

bvaradar opened the pull request:
Added documentation describing relevant terminologies and higher-level design and implementation details about Compaction.  Also added CLI section (both current and upcoming CLIs) for compaction.

bvaradar commented:
@n3nash : Can you take a first look at it to see if this is ok.

n3nash commented on docs/admin_guide.md:
```diff
@@ -158,6 +158,155 @@ hoodie:trips->stats wa
+    ____________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________
+    | Partition Path| File Id                             | Base Instant  | Data File Path                                                                                                                                                   | Total Delta Files| getMetrics                                                                                                                    |
+    |===========================================================================================================================================================================================================================================================================================================================================================================================|
+    | 2018/07/17    | 824a79ac-36d6-4d56-bd29-d182fe1a887d| 20180801184843| viewfs://ns-default/.../2018/07/17/824a79ac-36d6-4d56-bd29-d182fe1a887d_0_20180801184843.parquet | 1                | {TOTAL_LOG_FILES=1.0, TOTAL_IO_READ_MB=1230.0, TOTAL_LOG_FILES_SIZE=2.51255751E8, TOTAL_IO_WRITE_MB=991.0, TOTAL_IO_MB=2221.0}|
```
Can we add a dummy example here ? Like dummy committimes etc ?

n3nash commented on docs/admin_guide.md:
````diff
@@ -158,6 +158,155 @@ hoodie:trips->stats wa
+```
+
+```
+hoodie:trips->help compaction run
````
@bvaradar This is what you want me to add correct ?

n3nash commented on docs/admin_guide.md:
````diff
@@ -158,6 +158,155 @@ hoodie:trips->stats wa
+operations. NOTE: When running these operations, make sure no other concurrent compaction scheduling is happening
+
+```
+hoodie:trips->hoodie:schemaless_jobstore_orders_rows->help compaction schedule
````
Change the names of the tables, we can't leak out these names in OSS

n3nash commented on docs/admin_guide.md:
````diff
@@ -158,6 +158,155 @@ hoodie:trips->stats wa
+```
+
+To manually schedule or run a compaction, use the below command. This command uses spark launcher to perform compaction
+operations. NOTE: When running these operations, make sure no other concurrent compaction scheduling is happening
````
May be reword this. Say "NOTE : Make sure no other application is scheduling compaction for this dataset" ?

n3nash commented on docs/admin_guide.md:
````diff
@@ -158,6 +158,155 @@ hoodie:trips->stats wa
+
+```
+hoodie:trips->hoodie:schemaless_jobstore_orders_rows->help compaction schedule
+              Keyword:                   compaction schedule
````
Is this indentation right ? The one below is left indented and this one has tab indented

n3nash commented on docs/admin_guide.md:
```diff
@@ -158,6 +158,155 @@ hoodie:trips->stats wa
+In the next release, new useful CLI to revert/repair compaction schedules will be added. Here is a preview of them:
+
+
+To Validate a compaction plan. This CLI will check if all the files necessary for compactions are present and are valid
```
Change to : 

Validating a compaction plan : Check if all the files necessary for compactions are present and are valid

n3nash commented on docs/admin_guide.md:
```diff
@@ -158,6 +158,155 @@ hoodie:trips->stats wa
+
+##### Up-Coming CLI for Compaction
+
+In the next release, new useful CLI to revert/repair compaction schedules will be added. Here is a preview of them:
```
Change to : 

In the next release, more useful CLI to revert/repair compaction schedules will be added.

n3nash commented on docs/admin_guide.md:
````diff
@@ -158,6 +158,155 @@ hoodie:trips->stats wa
+hoodie:trips->compaction validate --compactionInstant <instantId>
+```
+
+All the below compaction commands needs to be run with both writer and ingestion application shut down:
````
Change to : 

The following commands must be executed without any other writer/ingestion application running.

n3nash commented on docs/concepts.md:
```diff
@@ -62,6 +62,24 @@ Each record is uniquely identified by a `record key` and mapped to a file id for
 and file id, never changes once the first version of a record has been written to a file. In short, the
  `file id` identifies a group of files, that contain all versions of a group of records.
 
+## Terminlogies
```
Terminologies

n3nash commented on docs/concepts.md:
```diff
@@ -62,6 +62,24 @@ Each record is uniquely identified by a `record key` and mapped to a file id for
+## Terminlogies
+
+ * `Hudi Dataset` 
+    A structured hive/spark table whose ingestion and querying is managed by Hudi. Hudi supports both partitioned and 
```
A structured hive/spark table managed by Hudi

n3nash commented on docs/concepts.md:
```diff
@@ -62,6 +62,24 @@ Each record is uniquely identified by a `record key` and mapped to a file id for
+    A structured hive/spark table whose ingestion and querying is managed by Hudi. Hudi supports both partitioned and 
+    non-partitioned Hive tables. 
+ * `Commit` 
+    A commit refers to one round of ingestion that brings in a batch of new (inserts) and updated records to a dataset.
```
A commit marks a new batch of data applied to a dataset..

n3nash commented on docs/concepts.md:
```diff
@@ -62,6 +62,24 @@ Each record is uniquely identified by a `record key` and mapped to a file id for
+    published.
+ * `Commit Timeline`
+    Commit Timeline refers to the sequence of Commits that was applied in order on a dataset over its lifetime. 
+ * `File Slice` 
```
FileSlice is a little more than just the version of a file id, but also includes any delta files and their versions, need a way to incorporate this in your text

n3nash commented on docs/implementation.md:
```diff
@@ -61,9 +63,157 @@ records such that
 #### Merge On Read
 
-Work in Progress .. .. .. .. ..
+In the Merge-On-Read storage model, the processes(spark application) for ingesting data (both inserts/updates) into the dataset
```
May be we don't want to say "processes" or the fact that they are decoupled because people may choose to run inline compactions or something similar, can we say that in this model we choose to defer merging/re-writing updates ?

n3nash commented on docs/implementation.md:
```diff
@@ -61,9 +63,157 @@ records such that
+ The key difference here is that updates are appended to latest log (delta) file belonging to the latest file slice 
+ without merging. For inserts, Hudi supports 2 modes:
+
+   1. Inserts To Log Files - For external indices which are capable of looking up record-keys that could have 
```
May be say : 

1. Inserts to Log Files - This is done for datasets that have an indexable log files (for eg global index)
2. Inserts to parquet files - This is done for datasets that do not have indexable log files, for eg bloom index embedded in parquer files...

n3nash commented on docs/implementation.md:
```diff
@@ -61,9 +63,157 @@ records such that
+  
+#### Compactor
+
+Realtime Readers will perform on-the-fly merge of these delta log-files to provide the most recent (committed) view of 
```
perform "in-situ" merge ....

n3nash commented on docs/implementation.md:
```diff
@@ -61,9 +63,157 @@ records such that
+#### Compactor
+
+Realtime Readers will perform on-the-fly merge of these delta log-files to provide the most recent (committed) view of 
+the dataset. To keep the query-performance in check, Hudi supports compacting these log-files asynchronously 
```
s/in check/unaffected or read-optimized..

n3nash commented on docs/implementation.md:
```diff
@@ -61,9 +63,157 @@ records such that
+
+Realtime Readers will perform on-the-fly merge of these delta log-files to provide the most recent (committed) view of 
+the dataset. To keep the query-performance in check, Hudi supports compacting these log-files asynchronously 
+(w.r.t writer)to create read-optimized views. Asynchronous Compaction involves 2 steps:
```
remove w.r.t writer..

n3nash commented on docs/implementation.md:
````diff
@@ -61,9 +63,157 @@ records such that
+     ```
+  * `Compactor` : Hudi provides a separate API in Write Client to execute a compaction plan. The compaction
+    plan (just like a commit) is identified by a timestamp. Most of the design and implementation complexity for Async 
+    Compaction has gone for guaranteeing snapshot isolation to readers and writer when
````
Change to : "Most of the design and implementation complexity for Async Compaction guarantees snapshot isolation to readers and writer.."

n3nash commented on docs/implementation.md:
```diff
@@ -61,9 +63,157 @@ records such that
+
+### Algorithm
+
+The algorithm is described with an illustration. Let us assume a scenario where there are commits SC1,DC2, DC3 that have 
```
DC1 instead of SC1

bvaradar replied:
It is supposed to be SC1

n3nash commented:
@vinothchandar LGTM

vinothchandar commented:
@bvaradar CI failures?

bvaradar commented:
@vinothchandar : Passes locally. The failure was because of memory allocation failure. Could be because of travis box issue ?

 I have rebased and updated this branch
